Introduce `DoNotMockDataClass` lint rule

wzieba opened the pull request:
<!-- Remember about a good descriptive title. -->

<!-- Id number of the Linear issue this PR addresses, e.g., WOOMOB-373. -->

Close: AINFRA-951
Internal: paqN3M-1w9-p2

### Description
<!-- Take the time to write a good summary. Why is it needed? What does it do? When fixing bugs try to avoid just writing “See original issue” – clarify what the problem was and how you’ve fixed it. -->

This PR introduces `DoNotMockDataClass` from Slack's lint (provided by WordPress lint wrapper). It also adds current violations to the baseline - I hope that in the future we'll address them.

And, on top of that, it improves configuration of Lint in this project. From now on:
- Custom lint checks are added to _all_ Library and App modules. If checks are not added, then rules from them don't run in a module. E.g., in this project "WordPress lint checks" were _not_ running in `fluxc-tests`. Now, they do.
- `WooCommerce-Wear`'s lint will now send sarif report to GitHub. This project will check only its own codebase.
- `WooCommerce`'s lint will check itself + all dependencies

The above is IMO optimal configuration:
- we benefit from performance of `checkDependencies=true` (see [Only analyze app/leaf modules](https://googlesamples.github.io/android-custom-lint-rules/user-guide.md.html))
- we don't duplicate violations from subprojects (`WooCommerce-Wear`'s lint e.g. doesn't include `FluxC`violations)

### Testing information
<!-- This is your opportunity to break out individual scenarios that need testing (when necessary) and/or include a checklist for the reviewer to go through. Consider documenting the following from your own completed testing: devices used, alternate workflows, edge cases, affected areas, critical flows, areas not tested, and any remaining unknowns. Provide feedback on this new section of the PR template through Sept 30, 2024 to Apps Quality; additional context here: p91TBi-b8z-p2#comment-12036 -->

See: https://github.com/woocommerce/woocommerce-android/pull/14288 . There are 3 violations:

- From `WooCommerce-Wear` - this means that sending `.sarif` from this module works 👍 
- From `fluxc-plugin` - this means that `checkDependencies` works fine from `WooCommerce` 👍 
- From `fluxc-test` - this is not detected as we use `abortOnError = true` (default). So whenever the violation from `fluxc-plugin` is detected, the Gradle job is canceled.

I was thinking about setting `abortOnError = false`, so all lint violations would be commented, but this also means that the Gradle task would return "success" every time (even if there are violations), and this doesn't sound good for dev experience.

### Screenshot

<img width="1036" height="102" alt="Screenshot 2025-07-17 at 16 35 39" src="https://github.com/user-attachments/assets/c7ce9f68-a192-41b4-b85b-d841c764c22f" />

- [ ] I have considered if this change warrants release notes and have added them to `RELEASE-NOTES.txt` if necessary. Use the "[Internal]" label for non-user-facing changes.

<!-- Pull request guidelines: https://github.com/woocommerce/woocommerce-android/blob/develop/docs/pull-request-guidelines.md -->

wzieba committed:
Introduce `DoNotMockDataClass` and `DoNotMockSealedClass` checks.

Baseline current issues.

wzieba committed:
Update `lint-baseline.xml` with a few missing cases

dangermattic commented:
<!--
  0 Errors
  1 Warning: This PR is larger than 300 lin...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is larger than 300 lines of changes. Please consider splitting it into smaller PRs for easier and faster reviews.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

wzieba committed:
Merge branch 'trunk' into do_not_mock_data_class_lint

wzieba committed:
Update WordPress Lint to `2.2.0`

wzieba committed:
Update lint baseline

wzieba committed:
Baseline only `DoNotMockDataClass` and `DoNotMockSealedClass` violations

wzieba committed:
Unify Lint configuration. Enable custom lint checks in subprojects

- apply custom lint checks to all AppPlugin and LibraryPlugin projects
- `WooCommerce`: when running lint, check subprojects as well
- `WooCommerce-Wear`: check only this module

This allows to keep clean and minimal lint baseline files.

wzieba committed:
Remove `InvalidPackage` config from `build.gradle`

It's handled in `lint.xml`

wzieba committed:
Do not fail fast on Lint tasks

I'd like lint to report all errors from all modules right away.

wzieba committed:
Upload `.sarif` report for WooCommerce-Wear as well

wzieba committed:
Set `checkDependencies=false` in WooCommerce-Wear instead of FluxC

It was a typo. This doesn't change much, as `checkDependencies` is `false` by default, but I want to be elaborative here that `WooCommerce-Wear` should only lint itself

wzieba committed:
Revert `abortOnError=false`

This was not failing the Gradle task in case of lint error.

ParaskP7 commented on .buildkite/commands/lint.sh:
```diff
@@ -20,5 +20,6 @@ if [ $app_lint_exit_code -ne 0 ] || [ $wear_lint_exit_code -ne 0 ]; then
 fi
 
 upload_sarif_to_github 'WooCommerce/build/reports/lint-results-jalapenoDebug.sarif'
+upload_sarif_to_github 'WooCommerce-Wear/build/reports/lint-results-jalapenoDebug.sarif'
```
Praise (❤️): Nice!

ParaskP7 commented on WooCommerce-Wear/build.gradle:
```diff
@@ -120,6 +120,10 @@ android {
     }
+
+    lintOptions {
+        checkDependencies = false
```
> [...but I want to be elaborative here that `WooCommerce-Wear` should only lint itself](https://github.com/woocommerce/woocommerce-android/pull/14287/commits/f8eaa4a76462261095154bf14c1d910474bc4618)

Praise (❤️): I agree with being explicit here!

ParaskP7 reviewed (approved):
👋 @wzieba !

I have reviewed and tested this PR as per the instructions, everything works as expected, nicely done, let's stop the mock bleeding! 🧪 x 🧪 ^ 🧪 

---

`EXTRAS`

> And, on top of that, it improves configuration of Lint in this project. From now on:

All great improvements! 🥇

> The above is IMO optimal configuration:

I agree! 👍

> I was thinking about setting abortOnError = false, so all lint violations would be commented, but this also means that the Gradle task would return "success" every time (even if there are violations), and this doesn't sound good for dev experience.

👍
